crates/chain-state/src/in_memory.rs

```diff
@@ -354,6 +356,35 @@ impl CanonicalInMemoryState {
     pub fn notify_canon_state(&self, event: CanonStateNotification) {
         self.inner.canon_state_notification_sender.send(event).ok();
     }
+
+    /// Return state provider with reference to in-memory blocks that overlay database state.
+    ///
+    /// This merges the state of all blocks that are part of the chain that the requested block is
+    /// the head of. This includes all blocks that connect back to the canonical block on disk.
+    pub fn state_provider(
+        &self,
+        hash: B256,
+        historical: StateProviderBox,
+    ) -> MemoryOverlayStateProvider {
+        let mut in_memory = Vec::new();
+        let mut current_hash = hash;
+
+        if let Some(first_state) = self.state_by_hash(hash) {
+            let anchor = first_state.anchor();
+
+            while let Some(state) = self.state_by_hash(current_hash) {
+                in_memory.insert(0, state.block());
+
+                if anchor.hash == current_hash {
+                    break;
+                }
+
+                current_hash = state.block().block().parent_hash;
+            }
+        }
+
+        MemoryOverlayStateProvider::new(in_memory, historical)
```

Comment:
I think we could actually simplify this a bit since the BlockState now has a list of all in memory parent blocks via the linked list:

https://github.com/paradigmxyz/reth/blob/c4bf5bb8ca271a6055f5af8431fd0bfb4ba79feb/crates/chain-state/src/in_memory.rs#L365-L366

so we can obtain the list of parents by traversing those nodes, and the hash that points to disk will be the anchor or parent hash of the last node in that list

Reply:
thanks for pointing that out! done here https://github.com/paradigmxyz/reth/pull/9817/commits/e277928d65bac03989f203f41e9dcdffe74cd8ab ptal